x/fee_grant/internal/types/tx/tx.go

```diff
@@ -0,0 +1,241 @@
+				signers = append(signers, addr)
+				seen[addr.String()] = true
+			}
+		}
+	}
+	return signers
+}
+
+// GetMemo returns the memo
+func (tx FeeGrantTx) GetMemo() string { return tx.Memo }
+
+// GetSignatures returns the signature of signers who signed the Msg.
+// CONTRACT: Length returned is same as length of
+// pubkeys returned from MsgKeySigners, and the order
+// matches.
+// CONTRACT: If the signature is missing (ie the Msg is
+// invalid), then the corresponding signature is
+// .Empty().
+func (tx FeeGrantTx) GetSignatures() [][]byte {
+	sigs := make([][]byte, len(tx.Signatures))
+	for i, stdSig := range tx.Signatures {
+		sigs[i] = stdSig.Signature
+	}
+	return sigs
+}
+
+// GetPubkeys returns the pubkeys of signers if the pubkey is included in the signature
+// If pubkey is not included in the signature, then nil is in the slice instead
+func (tx FeeGrantTx) GetPubKeys() []crypto.PubKey {
+	pks := make([]crypto.PubKey, len(tx.Signatures))
+	for i, stdSig := range tx.Signatures {
+		pks[i] = stdSig.PubKey
+	}
+	return pks
+}
+
+// GetSignBytes returns the signBytes of the tx for a given signer
+func (tx FeeGrantTx) GetSignBytes(ctx sdk.Context, acc exported.Account) []byte {
+	genesis := ctx.BlockHeight() == 0
+	chainID := ctx.ChainID()
+	var accNum uint64
+	if !genesis {
+		accNum = acc.GetAccountNumber()
+	}
+	return StdSignBytes(chainID, accNum, acc.GetSequence(), tx.Fee, tx.Msgs, tx.Memo)
+}
+
+// GetGas returns the Gas in GrantedFee
+func (tx FeeGrantTx) GetGas() uint64 { return tx.Fee.Gas }
+
+// GetFee returns the FeeAmount in GrantedFee
+func (tx FeeGrantTx) GetFee() sdk.Coins { return tx.Fee.Amount }
+
+// FeePayer returns the address that is responsible for paying fee
+// This can be explicily set in GrantedFee, or defaults to MainSigner
+func (tx FeeGrantTx) FeePayer() sdk.AccAddress {
+	if len(tx.Fee.FeeAccount) != 0 {
+		return tx.Fee.FeeAccount
+	}
+	return tx.MainSigner()
+}
+
+// MainSigner returns the first signer of the tx, by default this
+// account is responsible for fees, if not explicitly set.
+func (tx FeeGrantTx) MainSigner() sdk.AccAddress {
+	if len(tx.GetSigners()) != 0 {
+		return tx.GetSigners()[0]
+	}
+	return sdk.AccAddress{}
+}
+
+// GrantedFee includes the amount of coins paid in fees and the maximum
+// gas to be used by the transaction. The ratio yields an effective "gasprice",
+// which must be above some miminum to be accepted into the mempool.
+type GrantedFee struct {
+	Amount     sdk.Coins      `json:"amount" yaml:"amount"`
+	Gas        uint64         `json:"gas" yaml:"gas"`
+	FeeAccount sdk.AccAddress `json:"fee_account,omitempty" yaml:"fee_account"`
+}
+
+// NewGrantedFee returns a new instance of GrantedFee
+func NewGrantedFee(gas uint64, amount sdk.Coins, feeAccount sdk.AccAddress) GrantedFee {
+	return GrantedFee{
+		Amount:     amount,
+		Gas:        gas,
+		FeeAccount: feeAccount,
+	}
+}
+
+// Bytes for signing later
+func (fee GrantedFee) Bytes() []byte {
+	// normalize. XXX
```

Comment:
??

Reply:
See the source: https://github.com/cosmos/cosmos-sdk/blob/2222d0eb811066494da0758e416d200c9583639a/x/auth/types/stdtx.go#L174-L189

I would request that you do not hold these PRs to higher standards than the most core logic in the sdk - eg. bank and auth.

Reply:
I see, I thought it was left on purpose for a reason